backend/component/proxy.pm

```diff
@@ -0,0 +1,254 @@
+# the Free Software Foundation; either version 2 of the License, or
+# (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License along
+# with this program; if not, see <http://www.gnu.org/licenses/>.
+
+package backend::component::proxy;
+use base 'Mojolicious';
+use Mojo::Base 'backend::component::process';
+
+use Mojo::Server::Prefork;
+use Mojo::Server::Daemon;
+
+use Mojo::Transaction::HTTP;
+use bmwqemu;
+use constant PROXY_BASE_PORT => 9992;
+has listening_address        => "127.0.0.1";
+has listening_port           => PROXY_BASE_PORT;
+has redirect_table => sub { {} };
+has policy         => "FORWARD";
+has server_type    => "Prefork";
+has load => sub { $bmwqemu::vars{CONNECTIONS_HIJACK_PROXY} || $bmwqemu::vars{CONNECTIONS_HIJACK} };
+has silent_daemon      => 1;
+has log_level          => "error";
+has inactivity_timeout => 20;
+has max_redirects      => 5;
+has connect_timeout    => 20;
+has request_timeout    => 10;
+
+sub prepare {
+    my ($self) = @_;
+    my @entry;
+
+    my $hosts = $bmwqemu::vars{CONNECTIONS_HIJACK_PROXY_ENTRY};
+    my $policy = $bmwqemu::vars{CONNECTIONS_HIJACK_PROXY_POLICY} || $self->policy;    # Can be REDIRECT, DROP, FORWARD
+
+    $bmwqemu::vars{CONNECTIONS_HIJACK_PROXY_POLICY} = $policy
+      if !$bmwqemu::vars{CONNECTIONS_HIJACK_PROXY_POLICY} || $bmwqemu::vars{CONNECTIONS_HIJACK_PROXY_POLICY} ne $policy;
+
+    my $proxy_server_port
+      = $bmwqemu::vars{CONNECTIONS_HIJACK_PROXY_SERVER_PORT} || ($bmwqemu::vars{VNC} ? $bmwqemu::vars{VNC} + PROXY_BASE_PORT : $self->listening_port);
+    my $proxy_server_address = $bmwqemu::vars{CONNECTIONS_HIJACK_PROXY_SERVER_ADDRESS} || $self->listening_address;
+
+    $bmwqemu::vars{CONNECTIONS_HIJACK_PROXY_SERVER_PORT} = $proxy_server_port
+      if !$bmwqemu::vars{CONNECTIONS_HIJACK_PROXY_SERVER_PORT} || $bmwqemu::vars{CONNECTIONS_HIJACK_PROXY_SERVER_PORT} ne $proxy_server_port;
+
+    my $server_type = $bmwqemu::vars{CONNECTIONS_HIJACK_PROXY_SERVER_TYPE};
+
+    @entry = split(/,/, $hosts) if $hosts;
+
+    # Generate record table from configuration
+    my $redirect_table = {
+        map {
+            my ($host, @redirect) = split(/:/, $_);
+            $host => [@redirect] if ($host and @redirect)
+        } @entry
+    };
+
+    $policy = "REDIRECT" if (!$policy && $hosts);
+
+    # Handle SUSEMIRROR and MIRROR_HTTP transparently
+    if ($bmwqemu::vars{MIRROR_HTTP} || $bmwqemu::vars{SUSEMIRROR}) {
+        $self->_diag("Use of mirror detected. Setting up Proxy configuration automatically according to MIRROR_HTTP");
+
+        my $rules = [
+            ($bmwqemu::vars{MIRROR_HTTP} || $bmwqemu::vars{SUSEMIRROR}), "/tumbleweed/repo/.*oss/repodata",
```

Comment:
you will have to find a way to move all this into the distribution/tests - this won't go into os-autoinst